src/compiler/builder.ts

```diff
@@ -1098,6 +1102,33 @@ namespace ts {
             );
         }
 
+        function getWriteFileUpdatingSignatureCallback(writeFile: WriteFileCallback | undefined): WriteFileCallback {
+            return (fileName, text, writeByteOrderMark, onError, sourceFiles, data) => {
+                if (sourceFiles?.length === 1 && isDeclarationFileName(fileName)) {

```

Comment:
When is `sourceFiles.length === 1` true? When that’s true, does it also imply `fileName === sourceFiles[0].fileName`?

Reply:
> also imply `fileName === sourceFiles[0].fileName`?

No thats not true. The fileName = "d.ts" for sourceFile[0]
Its going to be single source file because its not `--out` options

Reply:
Let me see if I can try to parse out what you’re saying:

This callback writes one output file. `sourceFiles` is the set of input files that correspond to that one output file. This is always 1:1 (e.g. `foo.ts` when writing `foo.d.ts`) unless `outFile` is set, in which case all input files get emitted together into a single .js/.d.ts.

`sourceFiles?.length === 1 && isDeclarationFileName(fileName)` ensures we have a 1:1 input:output and that we have a reference to the input file for the .d.ts file being written. If there were no `sourceFiles` or if `fileName` was a JS file, we obviously wouldn’t be able to do the logic in the condition. If there were many source files, it would indicate that the signature for each file is all combined into one .d.ts and so it wouldn’t make sense to use that as an up-to-date check for any of them (an change in any file’s signature would affect all files).

Is that right?

Reply:
Yes